apps/aecore/src/aec_peers_pool.erl

```diff
@@ -0,0 +1,1695 @@
+%% If the peer is already pooled but its address changed, the update is ignored;
+%% this is so the pool cannot get poisoned by nodes gossiping wrong addresses
+%% for valid peers. If a peer rightfully changed its address, the old one will
+%% get removed after the standard retry procedure; at that point, the new one
+%% will be added. Note that this could take quite some time.
+%%
+%% If the peer is already in the verified pool, nothing is done beside updating
+%% the peer last update time and source address.
+%%
+%% If the peer is already in the unverified pool but has not reached the maximum
+%% number of references, another reference may be added to the unverified pool;
+%% the peer last update time and source address fields are updated too.
+%%
+%% Some opaque term can be specified; this term will be returned by `select/4`
+%% alongside the peer identifier. This could be used by the caller to store
+%% extra connection information like protocol and port number. This extra
+%% information will **NOT** be updated if the peer already exists.
+-spec update(state(), millitimestamp(), peer_id(),
+             peer_addr(), peer_addr(), boolean(), extra())
+  -> {verified | unverified | ignored, state()}.
+update(St, Now, PeerId, PeerAddr, SourceAddr, IsTrusted, Extra) ->
+  ?assertNotEqual(undefined, PeerId),
+  ?assertNotEqual(undefined, PeerAddr),
+  ?assertNotEqual(undefined, SourceAddr),
+  case update_peer(St, Now, PeerId, PeerAddr, SourceAddr, IsTrusted, Extra) of
+    ignored -> {ignored, St};
+    {updated, St2} ->
+      Result = case IsTrusted of
+        true -> verified_maybe_add(St2, Now, PeerId);
+        false -> unverified_maybe_add(St2, Now, PeerId, undefined)
+      end,
+      case Result of
+        {ignored, St3} ->
+          % Couldn't be added to any pool; must be removed.
+          {ignored, del_peer(St3, PeerId)};
+        _ -> Result
+      end
+  end.
+
+%% @doc Marks a peer as verified.
+%%
+%% Takes a time in millisecond as it would be returned from
+%% `erlang:system_time(millisecond)` and uses it as if it was the current time.
+%% Every calls should be given a time greater or equal to the last call.
+%%
+%% If the peer is in the unverified pool, it is added to the verified pool
+%% and all its references are removed from the unverified pool. If for some
+%% reason the insertion in the verified pool is not possible, the peer stays in
+%% the unverified pool.
+-spec verify(state(), millitimestamp(), peer_id())
+  -> {verified | ignored, state()}.
+verify(St, Now, PeerId) ->
+  verified_maybe_add(St, Now, PeerId).
+
+%% @doc Returns a random subset of all the pooled peers.
+%%
+%% Return a list of peer identifiers and there corresponding extra data given
+%% the **FIRST** time it was added by `update/7`.
+%%
+%% A filtering function can be specified to limit the possible results.
+-spec subset(state(), all | pos_integer(), ext_filter_fun() | undefined)
+  -> {[ext_peer()], state()}.
+subset(St, Size, ExtFilterFun) ->
```

Comment:
I feel the name could be clearer, like `random_subset` or `random_peers` or something.